### PR TITLE
Replace with new python tests

### PR DESCRIPTION
**Pull request by ProfFan**
Per #354

### CODE REVIEW

File: python/gtsam_py/python/examples/ImuFactorExample.py [hunk 1]
```diff
@@ -5,27 +5,32 @@
```
Comment (dellaert) [inline python/gtsam_py/python/examples/ImuFactorExample.py:5]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/examples/ImuFactorExample2.py [hunk 2]
```diff
@@ -8,20 +8,23 @@
...
-                   ConstantTwistScenario, ImuFactor, NonlinearFactorGraph,
-                   PinholeCameraCal3_S2, Point3, Pose3,
-                   PriorFactorConstantBias, PriorFactorPose3,
-                   PriorFactorVector, Rot3, Values)
-from gtsam import symbol_shorthand_B as B
-from gtsam import symbol_shorthand_V as V
-from gtsam import symbol_shorthand_X as X
 from mpl_toolkits.mplot3d import Axes3D  # pylint: disable=W0611
 
+import gtsam
```
Comment (dellaert) [inline python/gtsam_py/python/examples/ImuFactorExample2.py:15]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/examples/PlanarSLAMExample.py [hunk 3]
```diff
@@ -13,25 +13,24 @@
 
 from __future__ import print_function
 
-import gtsam
 import numpy as np
-from gtsam import symbol_shorthand_L as L
-from gtsam import symbol_shorthand_X as X
+
+import gtsam
```
Comment (dellaert) [inline python/gtsam_py/python/examples/PlanarSLAMExample.py:18]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/examples/Pose2SLAMExample_g2o.py [hunk 4]
```diff
@@ -82,7 +83,7 @@ def vector3(x, y, z):
     print ("Done!")
 
 if args.plot:
-    resultPoses = gtsam.utilities_extractPose2(result)
+    resultPoses = gtsam.extractPose2(result)
```
Comment (dellaert) [inline python/gtsam_py/python/examples/Pose2SLAMExample_g2o.py:86]: put extractPose2 be in utils?

File: python/gtsam_py/python/examples/Pose2SLAMExample_g2o.py [hunk 5]
```diff
@@ -53,15 +53,16 @@ def vector3(x, y, z):
 assert args.kernel == "none", "Supplied kernel type is not yet implemented"
 
 # Add prior on the pose having index (key) = 0
-priorModel = gtsam.noiseModel_Diagonal.Variances(vector3(1e-6, 1e-6, 1e-8))
-graph.add(gtsam.PriorFactorPose2(0, gtsam.Pose2(), priorModel))
+graphWithPrior = graph
```
Comment (dellaert) [inline python/gtsam_py/python/examples/Pose2SLAMExample_g2o.py:56]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/examples/Pose3SLAMExample_g2o.py [hunk 6]
```diff
@@ -39,21 +39,22 @@ def vector6(x, y, z, a, b, c):
 graph, initial = gtsam.readG2o(g2oFile, is3D)
 
 # Add Prior on the first key
-priorModel = gtsam.noiseModel_Diagonal.Variances(vector6(1e-6, 1e-6, 1e-6,
+priorModel = gtsam.noiseModel.Diagonal.Variances(vector6(1e-6, 1e-6, 1e-6,
```
Comment (dellaert) [inline python/gtsam_py/python/examples/Pose3SLAMExample_g2o.py:42]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/examples/Pose3SLAMExample_g2o.py [hunk 7]
```diff
@@ -65,7 +66,7 @@ def vector6(x, y, z, a, b, c):
     print ("Done!")
 
 if args.plot:
-    resultPoses = gtsam.utilities_allPose3s(result)
+    resultPoses = gtsam.allPose3s(result)
```
Comment (dellaert) [inline python/gtsam_py/python/examples/Pose3SLAMExample_g2o.py:69]: put in utils

File: python/gtsam_py/python/examples/PreintegrationExample.py [hunk 8]
```diff
@@ -111,7 +111,7 @@ def plotGroundTruthPose(self, t):
         actualPose = self.scenario.pose(t)
         plot_pose3(POSES_FIG, actualPose, 0.3)
         t = actualPose.translation()
-        self.maxDim = max([abs(t.x()), abs(t.y()), abs(t.z()), self.maxDim])
+        self.maxDim = max([max(t), self.maxDim])
```
Comment (dellaert) [inline python/gtsam_py/python/examples/PreintegrationExample.py:114]: port

File: python/gtsam_py/python/examples/README.md [hunk 9]
```diff
@@ -1,7 +1,8 @@
-These examples are almost identical to the old handwritten python wrapper
-examples. However, there are just some slight name changes, for example
-`noiseModel.Diagonal` becomes `noiseModel_Diagonal` etc...
-Also, instead of `gtsam.Symbol('b', 0)` we can simply say `gtsam.symbol_shorthand_B(0)` or `B(0)` if we use python aliasing.
+# THIS FILE IS OUTDATED!
```
Comment (dellaert) [inline python/gtsam_py/python/examples/README.md:1]: update it

File: python/gtsam_py/python/examples/SFMExample.py [hunk 10]
```diff
@@ -10,20 +10,24 @@
 """
 from __future__ import print_function
 
-import gtsam
 import matplotlib.pyplot as plt
 import numpy as np
-from gtsam import symbol_shorthand_L as L
-from gtsam import symbol_shorthand_X as X
+
+import gtsam
```
Comment (dellaert) [inline python/gtsam_py/python/examples/SFMExample.py:16]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/examples/SimpleRotation.py [hunk 11]
```diff
@@ -10,9 +10,8 @@
 a single variable with a single factor.
 """
 
-import gtsam
 import numpy as np
-from gtsam import symbol_shorthand_X as X
+import gtsam
```
Comment (dellaert) [inline python/gtsam_py/python/examples/SimpleRotation.py:14]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/examples/VisualISAM2Example.py [hunk 12]
```diff
@@ -13,15 +13,24 @@
...
-import gtsam
-import gtsam.utils.plot as gtsam_plot
 import matplotlib.pyplot as plt
 import numpy as np
-from gtsam import symbol_shorthand_L as L
-from gtsam import symbol_shorthand_X as X
-from gtsam.examples import SFMdata
 from mpl_toolkits.mplot3d import Axes3D  # pylint: disable=W0611
 
+import gtsam
```
Comment (dellaert) [inline python/gtsam_py/python/examples/VisualISAM2Example.py:20]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/examples/VisualISAMExample.py [hunk 13]
```diff
@@ -15,12 +15,15 @@
 import numpy as np
 import gtsam
 from gtsam.examples import SFMdata
-from gtsam.gtsam import (Cal3_S2, GenericProjectionFactorCal3_S2,
-                         NonlinearFactorGraph, NonlinearISAM, Point3, Pose3,
+from gtsam import (Cal3_S2, GenericProjectionFactorCal3_S2,
```
Comment (dellaert) [inline python/gtsam_py/python/examples/VisualISAMExample.py:18]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/tests/test_GaussianFactorGraph.py [hunk 14]
```diff
@@ -15,21 +15,20 @@
...
+import numpy as np
 
 def create_graph():
     """Create a basic linear factor graph for testing"""
     graph = gtsam.GaussianFactorGraph()
     
-    x0 = X(0)
-    x1 = X(1)
-    x2 = X(2)
+    x0 = gtsam.symbol('x', 0)
```
Comment (dellaert) [inline python/gtsam_py/python/tests/test_GaussianFactorGraph.py:26]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/tests/test_NonlinearOptimizer.py [hunk 15]
```diff
@@ -17,8 +17,7 @@
 import gtsam
 from gtsam import (DoglegOptimizer, DoglegParams, GaussNewtonOptimizer,
                    GaussNewtonParams, LevenbergMarquardtOptimizer,
-                   LevenbergMarquardtParams, PCGSolverParameters,
-                   DummyPreconditionerParameters, NonlinearFactorGraph, Ordering,
+                   LevenbergMarquardtParams, NonlinearFactorGraph, Ordering,
```
Comment (dellaert) [inline python/gtsam_py/python/tests/test_NonlinearOptimizer.py:20]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/tests/test_PriorFactor.py [hunk 16]
```diff
@@ -23,39 +23,17 @@ def test_PriorFactor(self):
 
         key = 5
         priorPose3 = gtsam.Pose3()
-        model = gtsam.noiseModel_Unit.Create(6)
+        model = gtsam.noiseModel.Unit.Create(6)
```
Comment (dellaert) [inline python/gtsam_py/python/tests/test_PriorFactor.py:26]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/tests/test_SFMExample.py [hunk 17]
```diff
@@ -34,29 +34,29 @@ def test_SFMExample(self):
         graph = gtsam.NonlinearFactorGraph()
 
         # Add factors for all measurements
-        measurementNoise = gtsam.noiseModel_Isotropic.Sigma(2, measurementNoiseSigma)
+        measurementNoise = gtsam.noiseModel.Isotropic.Sigma(2, measurementNoiseSigma)
```
Comment (dellaert) [inline python/gtsam_py/python/tests/test_SFMExample.py:37]: use shorthand here

File: python/gtsam_py/python/tests/test_Scenario.py [hunk 18]
```diff
@@ -43,8 +43,11 @@ def test_loop(self):
         # R = v/w, so test if loop crests at 2*R
         R = v / w
         T30 = scenario.pose(T)
+        xyz = T30.rotation().xyz()
```
Comment (dellaert) [inline python/gtsam_py/python/tests/test_Scenario.py:46]: what's this?

File: python/gtsam_py/python/tests/test_SimpleCamera.py [hunk 19]
```diff
@@ -5,7 +5,7 @@
 
 See LICENSE for the license information
 
-PinholeCameraCal3_S2 unit tests.
+SimpleCamera unit tests.
```
Comment (dellaert) [inline python/gtsam_py/python/tests/test_SimpleCamera.py:8]: SimpleCamera should no longer be there.

File: python/gtsam_py/python/utils/plot.py [hunk 20]
```diff
@@ -13,9 +13,8 @@ def set_axes_equal(fignum):
     Make axes of 3D plot have equal scale so that spheres appear as spheres,
     cubes as cubes, etc..  This is one possible solution to Matplotlib's
     ax.set_aspect('equal') and ax.axis('equal') not working for 3D.
-
-    Args:
-      fignum (int): An integer representing the figure number for Matplotlib.
+    Input
```
Comment (dellaert) [inline python/gtsam_py/python/utils/plot.py:16]: This script changed significantly and has to be re-ported

File: python/gtsam_py/python/utils/visual_data_generator.py [hunk 21]
```diff
@@ -1,17 +1,16 @@
 from __future__ import print_function
 
 import numpy as np
-
+from math import pi, cos, sin
```
Comment (dellaert) [inline python/gtsam_py/python/utils/visual_data_generator.py:4]: This script changed significantly and has to be re-ported

**Comment by dellaert** [comment]
PS @ProfFan I appreciate all the hard work - maybe @varunagrawal could port the IMU examples he changed significantly?

File: python/gtsam_py/python/examples/PlanarSLAMExample.py (hunk 3 above)
Comment (ProfFan) [reply python/gtsam_py/python/examples/PlanarSLAMExample.py:18]: I think this one is the same?

File: python/gtsam_py/python/examples/Pose2SLAMExample_g2o.py (hunk 5 above)
Comment (ProfFan) [reply python/gtsam_py/python/examples/Pose2SLAMExample_g2o.py:56]: this one is also virtually the same

File: python/gtsam_py/python/examples/Pose2SLAMExample_g2o.py (hunk 4 above)
Comment (ProfFan) [reply python/gtsam_py/python/examples/Pose2SLAMExample_g2o.py:86]: this is a typo

File: python/gtsam_py/python/examples/PlanarSLAMExample.py (hunk 3 above)
Comment (dellaert) [reply python/gtsam_py/python/examples/PlanarSLAMExample.py:18]: No. Uses shorthand now

File: python/gtsam_py/python/examples/Pose2SLAMExample_g2o.py (hunk 5 above)
Comment (dellaert) [reply python/gtsam_py/python/examples/Pose2SLAMExample_g2o.py:56]: No

**Comment by varunagrawal** [comment]
On it.

**Comment by ProfFan** [comment]
@varunagrawal Are you already working on it? if not I will start doing it?

File: python/gtsam_py/python/examples/PreintegrationExample.py (hunk 8 above)
Comment (ProfFan) [reply python/gtsam_py/python/examples/PreintegrationExample.py:114]: done

**Comment by varunagrawal** [comment]
I've started. :)

File: python/gtsam_py/python/examples/PlanarSLAMExample.py (hunk 3 above)
Comment (ProfFan) [reply python/gtsam_py/python/examples/PlanarSLAMExample.py:18]: done

File: python/gtsam_py/python/tests/test_SimpleCamera.py (hunk 19 above)
Comment (varunagrawal) [reply python/gtsam_py/python/tests/test_SimpleCamera.py:8]: @ProfFan @dellaert should this file just be removed then?

**Comment by varunagrawal** [comment]
@dellaert @ProfFan all done.

**Comment by ProfFan** [comment]
I tested all examples and tests, they all work perfectly now, thanks @varunagrawal !

**Comment by ProfFan** [comment]
@dellaert Two levels because we have other files that is non-python (not included in this PR)

**Comment by dellaert** [comment]
> @dellaert Two levels because we have other files that is non-python (not included in this PR)

Well, if they are related to wrapping, they should be in wrap. If they are not, then what are they? This is a major upgrade and we should do it right, this type of complexity lives with us for a long time.

**Comment by ProfFan** [comment]
@dellaert They are the necessary specializations for the aliased std types like `std::vector<xxx>`, whose types are not exposed by `gtsam.h` so not able to generate from the wrap script:
```cpp
PYBIND11_MAKE_OPAQUE(std::vector<gtsam::Key>);
PYBIND11_MAKE_OPAQUE(std::vector<gtsam::Point2, Eigen::aligned_allocator<gtsam::Point2> >);
PYBIND11_MAKE_OPAQUE(std::vector<gtsam::Pose3>);
PYBIND11_MAKE_OPAQUE(std::vector<gtsam::BetweenFactor<gtsam::Pose3>>);

py::bind_vector<std::vector<gtsam::Key> >(m_, "KeyVector");
py::bind_vector<std::vector<gtsam::Point2, Eigen::aligned_allocator<gtsam::Point2> > >(m_, "Point2Vector");
py::bind_vector<std::vector<gtsam::Pose3> >(m_, "Pose3Vector");
py::bind_vector<std::vector<boost::shared_ptr<gtsam::BetweenFactor<gtsam::Pose3> > > >(m_, "BetweenFactorPose3s");
```

**Comment by dellaert** [comment]
> @dellaert They are the necessary specializations for the aliased std types like `std::vector<xxx>`, whose types are not exposed by `gtsam.h` so not able to generate from the wrap script:
> 
> ```c++
> PYBIND11_MAKE_OPAQUE(std::vector<gtsam::Key>);
> PYBIND11_MAKE_OPAQUE(std::vector<gtsam::Point2, Eigen::aligned_allocator<gtsam::Point2> >);
> PYBIND11_MAKE_OPAQUE(std::vector<gtsam::Pose3>);
> PYBIND11_MAKE_OPAQUE(std::vector<gtsam::BetweenFactor<gtsam::Pose3>>);
> 
> py::bind_vector<std::vector<gtsam::Key> >(m_, "KeyVector");
> py::bind_vector<std::vector<gtsam::Point2, Eigen::aligned_allocator<gtsam::Point2> > >(m_, "Point2Vector");
> py::bind_vector<std::vector<gtsam::Pose3> >(m_, "Pose3Vector");
> py::bind_vector<std::vector<boost::shared_ptr<gtsam::BetweenFactor<gtsam::Pose3> > > >(m_, "BetweenFactorPose3s");
> ```

How many files are we talking about? Does not seem incompatible with eliminating the gtsam_py/python hierarchy...

**Comment by ProfFan** [comment]
@dellaert  This is to establish a standard for wrapping any library with the utility:

```
<top library>
|- src
|- include
|- python
|  | - <first_python_pkg>
|  |   | - python # python top module wrapper
|  |   |   | - __init__.py
|  |   |   | - utils
|  |   | - preamble.h # MAKE_OPAQUE declarations
|  |   | - specializations.h # std::vector specializations
|  | - <second_python_pkg>
|  |   |  .......
```

Probably `python` is a bad name, probably name it `top` or something?

**Comment by ProfFan** [comment]
@dellaert This is the current directory structure for GTSAM:
image

**Comment by dellaert** [comment]
Ideally, we would have:

```
python
  gtsam
    tests
    examples
  gtsam_unstable
    tests
    examples
```

Maybe you and I should chat - but tomorrow.

**Comment by ProfFan** [comment]
@dellaert  Sure :)

**Comment by ProfFan** [comment]
@dellaert I think that can be postponed to some time after we get develop to use pybind.

**Comment by dellaert** [comment]
> @dellaert I think that can be postponed to some time after we get develop to use pybind.

I disagree. The PR review is the right place to fix these things.

**Comment by ProfFan** [comment]
but then where can we put these non-python files? they are different for each wrapped module.

**Comment by dellaert** [comment]
I see. So, that was something nice of having an install directory -> we can access the tests/examples etc without knowing where they are in source/build. Maybe we should go back to that? I'm regretting we fast-tracked that change.

Let's talk tomorrow?

**Comment by ProfFan** [comment]
@dellaert Sure! I am available whole day, just send me an invite.

**Comment by ProfFan** [comment]
@dellaert Note that the tests are symlinked to the build folder, so you can also run tests from the usual path without knowing it's actual location.

File: python/gtsam_py/python/utils/visual_data_generator.py [hunk 22]
```diff
@@ -1,17 +1,17 @@
...
```
Comment (dellaert) [inline python/gtsam_py/python/utils/visual_data_generator.py:4]: Add 2.7 to see what the work is... I vote for test being dual-support, but example be 3.X -> then remove all __future__ in the examples.

File: python/gtsam_py/python/utils/visual_data_generator.py [hunk 23]
```diff
@@ -1,17 +1,17 @@
...
+from gtsam import PinholeCameraCal3_S2
 
 
 class Options:
     """
     Options to generate test scenario
     """
 
-    def __init__(self, triangle=False, nrCameras=3, K=Cal3_S2()):
+    def __init__(self, triangle=False, nrCameras=3, K=gtsam.Cal3_S2()):
```
Comment (dellaert) [inline python/gtsam_py/python/utils/visual_data_generator.py:14]: import classes used more than once

File: python/gtsam_py/python/examples/ImuFactorISAM2Example.py [hunk 24]
```diff
@@ -1,6 +1,6 @@
 """
 iSAM2 example with ImuFactor.
-Author: Robert Truax (C++), Frank Dellaert (Python)
+Author: Frank Dellaert, Varun Agrawal
```
[the remark this answers is not in the document]
Comment (varunagrawal) [reply python/gtsam_py/python/examples/ImuFactorISAM2Example.py:3]: His name is in the C++ version of the script.

[the remark this answers is not in the document]
Comment (dellaert) [reply python/gtsam_py/python/examples/ImuFactorISAM2Example.py:3]: Just add it back :-)

**Comment by dellaert** [comment]
@ProfFan let me know when changes have been made so we can merge

File: python/gtsam_py/python/examples/ImuFactorISAM2Example.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (ProfFan) [reply python/gtsam_py/python/examples/ImuFactorISAM2Example.py]: I'll add it

**Comment by ProfFan** [comment]
Done

File: python/gtsam_py/python/examples/ImuFactorISAM2Example.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (varunagrawal) [reply python/gtsam_py/python/examples/ImuFactorISAM2Example.py]: Thanks! :)

**Pull request merged**